keps/sig-auth/0000-authority-delegation.md

```diff
@@ -0,0 +1,259 @@
+  can implement as far as audit logging and access transparency. Tracing the
+  authority of a low level action to the source is painful, imprecise,
+  dependent on k8s version, even when each actor in the path is cooperative.
+
+The proposed authorization model seeks to solve these issues.
+
+<!---
+
+## Table of Contents
+
+## Summary
+
+
+### Goals
+
+### Non-Goals
+
+-->
+
+
+## Proposal
+
+### Selected Authority
+
+Today Kubernetes supports a kind of ambient authority. However support for
+selected authority is a generally useful model to support.
+
+This is particularly useful in a deputy scenario to prevent confused deputies.
+When a controller acts on behalf of an end principal, the authority by which it
+acts should be explicit. This prevents a controller from misusing permission
+granted ambiently to the controller.
+
+Authority is selected per request by passing an “Authority-Selector” header
+with the request:
+
+```
+Authority-Selector: {“authorityReference”:{“kind”:”ClusterRole”,”name”:”viewer”}}
+```
+or:
+```
+Authority-Selector: {“authorityReference”:{“kind”:”AuthorityDelegation”,”uid”:”0000-0...”}}
+```
+
+The value of the selector is a `SelectedAuthority` object in the
+`authorization.k8s.io` API group:
+
+```go
+type SelectedAuthority struct {
+  metav1.TypeMeta
+  AuthorityReference AuthorityReference
+}
+
+type AuthorityReference struct {
+  metav1.TypeMeta
+  Name string
+  UID types.UID
+}
+```
+
+When a request is received by an API, the API would normally iterate over the
+set of authorizers and each authorizer would check all policies that would
+permit the action, searching for a policy that permits the action and is bound
+to the user.
+
+When a request is received by the API with a selected authority, the API would
+check the specific authorizer that supports the AuthorityReference kind, and
+the authorizer would check the request against the specific policy referenced
+in by the SelectedAuthority.
+
+### Persistent Authority Delegation
+
+#### AuthorityDelegation
+
+A new non-namespaced persistent object will be added to the
+`authorization.k8s.io` API group to store instances of authority delegation.
+
+```go
+type AuthoritiyDelegation struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // The user that created the delegation.
+  User Info
+  // The principal that is delegating authority.
+  Subject SubjectRef
+  // The principals that are delegated authority.
+  Delegates []SubjectRef
+  // defined above
+  Authority AuthorityReference
+  // If this delegation is created selecting a previously delegated authority,
+  // the delegation chain is recorded here.
+  DelegationHistory []AuthorityDelegationReference
+}
+
+type SubjectRef struct {
+  Kind SubjectKind
+  Name string
+  UID types.UID
+}
+
+const (
+  GroupKind SubjectKind = "Group"
+  UserKind SubjectKind = "User"
+)
+
+type AuthrityDelegationReference struct {
+  UID types.UID
```

Comment:
Can we also possibly have the name of the authority here as well? More or less for human user _usage_